chain/chain/src/resharding.rs

```diff
@@ -517,4 +574,77 @@ impl Chain {
         }
         Ok(())
     }
+
+    /// Preprocessing stage for on demand resharding. This method should be called only by the resharding tool.
+    pub fn custom_build_state_for_resharding_preprocessing(
+        &self,
+        // The resharding will execute on the post state of the target_hash block.
+        target_hash: &CryptoHash,
+        // The block hash of the state snapshot.
+        snapshot_hash: &CryptoHash,
+        shard_id: ShardId,
+    ) -> Result<ReshardingRequest, Error> {
+        let sync_hash = target_hash;
+        tracing::debug!(target: "resharding", ?shard_id, ?sync_hash, "preprocessing started");
+        let block_header = self.get_block_header(sync_hash)?;
+        let shard_layout = self.epoch_manager.get_shard_layout(block_header.epoch_id())?;
+        let next_epoch_id = block_header.next_epoch_id();
+        let next_epoch_shard_layout = self.epoch_manager.get_shard_layout(next_epoch_id)?;
+        assert_ne!(shard_layout, next_epoch_shard_layout);
+
+        let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+        let state_root = *self.get_chunk_extra(&target_hash, &shard_uid)?.state_root();
+
+        let resharding_request = ReshardingRequest {
+            tries: Arc::new(self.runtime_adapter.get_tries()),
+            sync_hash: *sync_hash,
+            prev_hash: *target_hash,
+            prev_prev_hash: *snapshot_hash,
+            shard_uid,
+            state_root,
+            next_epoch_shard_layout,
+            curr_poll_time: Duration::ZERO,
+            config: self.resharding_config.clone(),
+            handle: self.resharding_handle.clone(),
+            on_demand: true,
+        };
+
+        // Technically it's not scheduled yet but it should be immediately after
+        // this method returns.
+        RESHARDING_STATUS
+            .with_label_values(&[&shard_uid.to_string()])
+            .set(ReshardingStatus::Scheduled.into());
+
+        Ok(resharding_request)
+    }
+
+    /// Postprocessing stage for on demand resharding. This method should be called only by the resharding tool.
+    pub fn custom_build_state_for_split_shards_postprocessing(
+        &mut self,
+        shard_uid: ShardUId,
+        sync_hash: &CryptoHash,
+        state_roots: HashMap<ShardUId, StateRoot>,
+    ) -> Result<(), Error> {
+        let block_header = self.get_block_header(sync_hash)?;
+        let prev_hash = block_header.prev_hash();
+
+        for (shard_uid, state_root) in state_roots {
+            debug!(target:"resharding", "Finish building resharding for shard {:?} {:?} {:?} ", shard_uid, prev_hash, state_root);
+            // Chunk extra should match the ones committed into the chain.
+            let new_chunk_extra = ChunkExtra::new_with_only_state_root(&state_root);
+            if let Ok(chunk_extra) = self.get_chunk_extra(sync_hash, &shard_uid) {
```

Comment:
This check has failed on me because the `chunk_extra` are missing. It looks like `build_state_for_split_shards_postprocessing` was never called before. I would be ok with removing this part of the code.

Reply:
Hmm this check is important but I wonder how it can be done. What error are you getting here? Can you figure out on what state exactly are we resharding and whether this chunk extra is expected to be present on chain? The sensible options would be [pre/post] state x [last block of old layout, first block of new layout].

Reply:
It's possible (worth checking) that the cold loop does not correctly copy chunk extra from hot to cold in which case this info would be missing. If this is the case perhaps you can compare it with the prev state root of the next chunk?

Reply:
I'm getting a `chunk extra not found` error. The chunk extra should be persisted by `build_state_for_split_shards_postprocessing` with 
```
chain_store_update.save_chunk_extra(&prev_hash, &shard_uid, chunk_extra);`
```
I have other `chunk_extra` in the cold db, but this one is missing. Maybe I can persisting it now instead of checking, not sure we again anything though.

I'll check next if there's some mismatch in the hash (one block later or before) since we do things a bit differently